distribution/xfer/download.go

```diff
@@ -414,11 +418,15 @@ func (ldm *LayerDownloadManager) makeDownloadFuncFromDownload(descriptor Downloa
 			}
 			defer layerReader.Close()
 
-			var src *distribution.Descriptor
-			if fs, ok := l.(layer.ForeignSourcer); ok {
```

Comment:
@aaronlehmann trying to combine both of these sections into a function as you suggested. I have my doubts over the correctness of this logic here. It uses `l` instead of `descriptor` like above. The concern I see when using `l` is that it represents a different layer download of the same data. The `key` used to store the downloads does not take into account URLs, therefore there could be a situation where the URLs are either unintentionally received from the other download or the URLs for this layer are ignored. This problem could be more general since a difference in URLs will not be represented in the layer id, if the parent happens to be the same. Regardless of that issue though, from this point in the code the original descriptor should be used. WDYT?